docs/zk_rollup_after_eip4844.md

```diff
@@ -0,0 +1,96 @@
+- [Data Availability Post 4844](https://scroll.io/blog/data-availability-4844)
+
+## Introduction
+
+**EIP-4844** (or Proto-Danksharding) is the backbone of Ethereum Dencun upgrade. This upgrade focuses on Layer 2 scalability, by
+introducing blob transactions and blob data, storage costs and gas fees have been significantly reduced.
+
+Here are a few important points of EIP-4844 from ZK-Rollup's perspective:
+
+- Data now can be stored in blob, which contains $4096$ chunks of $32$ bytes each.
+- The execution layer (can be understood as in smart contract) only can access
+  to [versioned hash](https://eips.ethereum.org/EIPS/eip-4844#helpers)
+  of [KZG commitment](../terms/polynomial-commitment/100_kate_commitment.md) of data in blob.
+- [BLS12-381](../terms/bls12-381.md) modulus is used in KZG commitment.
+
+## KZG commitments of blob data
+
+The function $f$ is defined as the interpolation polynomial:
+
+$$
+f(\omega^i) = d_i
+$$
+
+where $\omega^{4096} = 1$ is a [root of unity](plonk.md#Roots%20of%20Unity) of order $4096$, and $d_i$ define the data points in blob.
+This function is committed using KZG.
+
+## Point Evaluation Precompile
+
+The EIP4844 introduces a new precompile that is designed to allow users to open the commitment to the blobs, and effectively access the
+blob data from smart contracts. This is very handy when verifying optimistic or succinct proofs that involve the blob data. Below is
+the pseudocode copied from [EIP-4844 specs](https://eips.ethereum.org/EIPS/eip-4844).
+
+```python
+def point_evaluation_precompile(input: Bytes) -> Bytes:
+    """
+    Verify p(z) = y given commitment that corresponds to the polynomial p(x) and a KZG proof.
+    Also verify that the provided commitment matches the provided versioned_hash.
+    """
+    # The data is encoded as follows: versioned_hash | z | y | commitment | proof | with z and y being padded 32 byte big endian values
+    assert len(input) == 192
+    versioned_hash = input[:32]
+    z = input[32:64]
+    y = input[64:96]
+    commitment = input[96:144]
+    proof = input[144:192]
+
+    # Verify commitment matches versioned_hash
+    assert kzg_to_versioned_hash(commitment) == versioned_hash
+
+    # Verify KZG proof with z and y in big endian format
+    assert verify_kzg_proof(commitment, z, y, proof)
+
+    # Return FIELD_ELEMENTS_PER_BLOB and BLS_MODULUS as padded 32 byte big endian values
+    return Bytes(U256(FIELD_ELEMENTS_PER_BLOB).to_be_bytes32() + U256(BLS_MODULUS).to_be_bytes32())
+```
+
+The `point_evaluation_precompile(versioned_hash, kzg_commitment, proof, z, y)` receives a versioned hash, the KZG commitment to the
+blob, and a KZG opening proof for point $z$ and value $y$ as input. It verifies that `kzg_commitment` corresponds to
+the `versioned_hash` provided and that the opening `proof` is valid. That is the for Lagrange interpolation polynomial of the
+blob `p(X)` we have `p(z) = y`.
+
+## Blob Consistency Check
```

Comment:
How does this look like when implementing? Use the links provided in the general comment